tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1305 @@
+    ///
+    ///     let mut buf = [0u8; 4];
+    ///     client.read_exact(&mut buf).await?;
+    ///
+    ///     Ok::<_, io::Error>(buf)
+    /// });
+    ///
+    /// let mut buf = [0u8; 4];
+    /// let mut available = 0;
+    ///
+    /// for n in 0..N {
+    ///     if available < 4 {
+    ///         server.read_exact(&mut buf).await?;
+    ///         assert_eq!(&buf[..], b"ping");
+    ///
+    ///         let info = server.peek(None)?;
+    ///         available = info.total_bytes_available;
+    ///         continue;
+    ///     }
+    ///
+    ///     // here we know that at least `available` bytes are immediately
+    ///     // ready to read.
+    ///     server.read_exact(&mut buf).await?;
+    ///     available -= buf.len();
+    ///     assert_eq!(&buf[..], b"ping");
+    /// }
+    ///
+    /// server.write_all(b"pong").await?;
+    ///
+    /// let buf = client.await??;
+    /// assert_eq!(&buf[..], b"pong");
+    /// # Ok(()) }
+    /// ```
+    pub fn peek(&mut self, mut buf: Option<&mut ReadBuf<'_>>) -> io::Result<PipePeekInfo> {
+        use std::convert::TryFrom as _;
+
+        unsafe {
+            let mut n = mem::MaybeUninit::zeroed();
+            let mut total_bytes_available = mem::MaybeUninit::zeroed();
+            let mut bytes_left_this_message = mem::MaybeUninit::zeroed();
+
+            let result = {
+                let (buf, len) = match &mut buf {
+                    Some(buf) => {
+                        let len = DWORD::try_from(buf.capacity())
+                            .expect("buffer too large for win32 api");
+                        // Safety: the OS has no expectation on whether the
+                        // buffer is initialized or not.
+                        let buf = buf.inner_mut() as *mut _ as *mut _;
+                        (buf, len)
+                    }
+                    None => (ptr::null_mut(), 0),
+                };
+
+                namedpipeapi::PeekNamedPipe(
+                    self.io.as_raw_handle(),
+                    buf,
+                    len,
+                    n.as_mut_ptr(),
+                    total_bytes_available.as_mut_ptr(),
+                    bytes_left_this_message.as_mut_ptr(),
+                )
+            };
+
+            if result == FALSE {
+                return Err(io::Error::last_os_error());
+            }
+
+            let n = usize::try_from(n.assume_init()).expect("output size too large");
+
+            if let Some(buf) = buf {
+                // Safety: we trust that the OS has initialized up until `n`
+                // through the call to PeekNamedPipe.
+                buf.assume_init(n);
+            }
+
+            let total_bytes_available = usize::try_from(total_bytes_available.assume_init())
+                .expect("available bytes too large");
+            let bytes_left_this_message = usize::try_from(bytes_left_this_message.assume_init())
+                .expect("bytes left in message too large");
```

Comment:
Should these be panics?

Reply:
I'm not sure how you'd handle an error that told you that your platform-specific numerical types are too small, so probably *yes* (but I appreciate that [there is disagreements in implementation](https://github.com/yoshuawuyts/miow/blob/a84a1c0a3e96d48f89d3468f86d8dafc06b56268/src/handle.rs#L111)).

And at least the latter can't be clamped, because it's required to report the exact size of a message frame, which if incorrect could lead to logic errors.